*: skip flaky tests only optionally

tbg opened the pull request:
Using `t.Skip{f,}` as the weapon of choice for flaky tests is not a good
solution as tests which never run tend to rot and wind up not only flaky, but
completely broken (see #10141).

On the other hand, _not_ skipping flaky tests is also doomed to fail since the
toll on developers is too high.

Introduce an alternative which skips tests based on the environment variable
`COCKROACH_SKIP_FLAKY_TESTS` whenever `flaky.Register(<issue_num>, <reason>)`
is called, along with a simple lint to avoid accidental use of `t.Skip{,f}`.
With the environment variable set to a truthy value, the test is skipped with
an informative message. Otherwise, a similarly informative message is logged as
a `Warning`, but the test continues to run.

Change existing `Skip{f,}`s to use this method, fixing up some tests in the
process and documenting completely broken ones (through an extra `Skip`).

Change the TeamCity build scripts to set `COCKROACH_SKIP_FLAKY_TESTS=true`.

The intention is that henceforth, we can more liberally mark tests as flaky and
still not have them rot as they still run occasionally, though unsatisfyingly
mostly on developer's machines. There's also a small upshot to stress testing:
there's no need to unskip a test which is to be stressed.

NB: Using a command-line parameter was considered, but the need to have that
flag registered with every package made it the less palatable option.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10197)

<!-- Reviewable:end -->

tbg committed:
rsg: unskip tests which are "broken on CircleCI"

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I suspect @petermattis and @spencerkimball may have dissenting opinions, though. @andreimatei may as well.

---

Reviewed 1 of 1 files at r1, 23 of 23 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/cli_test.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvWesu7QpgvJjk5sej:-KUvWesu7QpgvJjk5sek:b-ujd7te) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/acceptance/cli_test.go#L39)):*
> ```Go
> func TestDockerCLI(t *testing.T) {
> 	if err := testDockerOneShot(t, "cli_test", []string{"stat", cluster.CockroachBinaryInContainer}); err != nil {
> 		flaky.Register(t, 6086, "no binary in one-shot container", err.Error())
> ```

this is actually not flaky, it's something different

---

*[pkg/acceptance/reference_test.go, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvWr5om_Kl_BXq4HXw:-KUvWr5om_Kl_BXq4HXx:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/acceptance/reference_test.go#L29)):*
> ```Go
> func runReferenceTestWithScript(t *testing.T, script string) {
> 	if err := testDockerOneShot(t, "reference", []string{"stat", cluster.CockroachBinaryInContainer}); err != nil {
> 		flaky.Register(t, 6086, "no binary in one-shot container", err.Error())
> ```

ditto

---

*[pkg/sql/copy_in_test.go, line 296 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvX92nEkgRRdB5VliK:-KUvX92nEkgRRdB5VliL:bttvbyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/sql/copy_in_test.go#L296)):*
> ```Go
> 
> 	t.Skip("since it was marked as flaky due to race detector warnings, this test has completely rotted and now times out")
> 	flaky.Register(t, 0, "https://github.com/lib/pq/issues/494")
> ```

cc @mjibson 

---

*[pkg/storage/client_metrics_test.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXFgU2Z1OjJC8jHVE:-KUvXFgU2Z1OjJC8jHVF:b4p3ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_metrics_test.go#L152)):*
> ```Go
> 		{m.RdbBlockCacheUsage, 0},
> 		{m.RdbBlockCachePinnedUsage, 0},
> 		{m.RdbBloomFilterPrefixChecked, 0}, // see #10085
> ```

did you mean to include these?

---

*[pkg/storage/client_raft_test.go, line 2089 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXUgHqu8J1vBPi69m:-KUvXUgHqu8J1vBPi69n:b6hajgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_raft_test.go#L2089)):*
> ```Go
> 	t.Skip("needs update for change to raft transport error reporting")
> 	// It's also flaky.
> 	flaky.Register(t, 8303)
> ```

this number changed?

---

*[pkg/util/flaky/flaky.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXzrvwk-3fto4LbbF:-KUvXzrvwk-3fto4LbbG:b-iwr2j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/util/flaky/flaky.go#L27)):*
> ```Go
> )
> 
> // Skipper is a narrow subset of testing.TB required for Register().
> ```

any reason not to use `testing.TB` directly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-:-KUvYDoWgzjMq1HRtbGC:bgtqqw5)*
<!-- Sent from Reviewable.io -->

jordanlewis commented on build/teamcity-acceptance.sh:
```diff
@@ -1,5 +1,8 @@
 #!/usr/bin/env bash
 set -euxo pipefail
+
+export COCKROACH_SKIP_FLAKY_TESTS=true
```
Unfortunately, this won't work - `builder.sh` only passes a particular set of environment variables through to the Docker environment. You'll need to add `COCKROACH_SKIP_FLAKY_TESTS` to the list, or just change `builder.sh` to pass through all of `env`.

jordanlewis commented on build/teamcity-stress.sh:
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/env bash
 set -euxo pipefail
 
+export COCKROACH_SKIP_FLAKY_TESTS=true
```
ditto

jordanlewis commented on build/teamcity-test.sh:
```diff
@@ -1,3 +1,6 @@
 #!/usr/bin/env bash
 set -euxo pipefail
+
+export COCKROACH_SKIP_FLAKY_TESTS=true
```
ditto

jordanlewis commented on build/teamcity-testrace.sh:
```diff
@@ -1,3 +1,6 @@
 #!/usr/bin/env bash
 set -euxo pipefail
+
+export COCKROACH_SKIP_FLAKY_TESTS=true
```
ditto

andreimatei commented:
I don't feel very strongly about this, but I'd like to hear more about the purpose.
Seems to me that the problem is not that we don't run flaky tests, is that they stay skipped for too long. And I guess you're trying to address this by having them run on developer's machines, so that they'll feel some pain and ping the respective bugs, and that would prod the assigned person to fix them.
Does that sum it? If so, how about we create some mechanism that more directly prods people to fix them. E.g. automated emails, or 30s in the team meeting where the people with the most/oldest skipped tests are publicly shamed? These would not have the downside that... running flaky tests is annoying when they fail on your unrelated change.

RaduBerinde commented:
I'm not sure about the policy either. I understand the problem, but should annoying all devs really be the preferred way to get these bugs fixed?

An alternative: make the new infrastructure allow running *just* the skipped tests, and set up a daily stress run with just these tests with a report sent out (preferably with rate of failure for each test).

Somewhat related to this - sooner or later we will have to have two tiers of tests, with the second tier having tests that are too slow. We could easily turn up some of the existing tests (e.g. the sql parallel test) to find more bugs, and we will probably want to add more stress tests (e.g. for dist sql). Just bringing it up because it may make sense to think about how this would work as well. Automated special daily runs of the "slow" tests might be the best bet here too.

tamird commented:
You can also use `env` to pass stuff you want. See `teamcity-stress.sh`.

On Oct 25, 2016 09:32, "Jordan Lewis" <notifications@github.com> wrote:

> *@jordanlewis* requested changes on this pull request.
> ------------------------------
>
> In build/teamcity-acceptance.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,5 +1,8 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> Unfortunately, this won't work - builder.sh only passes a particular set
> of environment variables through to the Docker environment. You'll need to
> add COCKROACH_SKIP_FLAKY_TESTS to the list, or just change builder.sh to
> pass through all of env.
> ------------------------------
>
> In build/teamcity-stress.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,6 +1,8 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
>
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
> ------------------------------
>
> In build/teamcity-test.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,3 +1,6 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
> ------------------------------
>
> In build/teamcity-testrace.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,3 +1,6 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPFPzXoTqqj0LJIY5elGZECB6yUW6ks5q3gTmgaJpZM4Kfv2e>
> .
>

tamird commented:
Yes, automated emails and public shaming would both be good. Remember that
you can set this environment variable in your shell, if you really desire
to avoid this discomfort.

On Tue, Oct 25, 2016 at 11:36 AM, Andrei Matei <notifications@github.com>
wrote:

> I don't feel very strongly about this, but I'd like to hear more about the
> purpose.
> Seems to me that the problem is not that we don't run flaky tests, is that
> they stay skipped for too long. And I guess you're trying to address this
> by having them run on developer's machines, so that they'll feel some pain
> and ping the respective bugs, and that would prod the assigned person to
> fix them.
> Does that sum it? If so, how about we create some mechanism that more
> directly prods people to fix them. E.g. automated emails, or 30s in the
> team meeting where the people with the most/oldest skipped tests are
> publicly shamed? These would not have the downside that... running flaky
> tests is annoying when they fail on your unrelated change.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10197#issuecomment-256071902>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPDgx5QPsilpr3QPfCMgX-U55Fa1lks5q3iIYgaJpZM4Kfv2e>
> .
>

vivekmenezes commented:
Good call Tobi. Nicely done

On Tue, Oct 25, 2016, 9:32 AM Jordan Lewis <notifications@github.com> wrote:

> *@jordanlewis* requested changes on this pull request.
> ------------------------------
>
> In build/teamcity-acceptance.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,5 +1,8 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> Unfortunately, this won't work - builder.sh only passes a particular set
> of environment variables through to the Docker environment. You'll need to
> add COCKROACH_SKIP_FLAKY_TESTS to the list, or just change builder.sh to
> pass through all of env.
> ------------------------------
>
> In build/teamcity-stress.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,6 +1,8 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
>
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
> ------------------------------
>
> In build/teamcity-test.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,3 +1,6 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
> ------------------------------
>
> In build/teamcity-testrace.sh
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>
> :
>
> > @@ -1,3 +1,6 @@
>  #!/usr/bin/env bash
>  set -euxo pipefail
> +
> +export COCKROACH_SKIP_FLAKY_TESTS=true
>
> ditto
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10197#pullrequestreview-5649065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBCzjo3rLYJW5v00Xb4roR_wIi0QPks5q3gTwgaJpZM4Kfv2e>
> .
>

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[pkg/util/flaky/flaky.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUxDfpi43Jvnoe_m0wh:-KUxDfpi43Jvnoe_m0wi:b-hzae41) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/util/flaky/flaky.go#L32)):*
> ```Go
> }
> 
> const envVarName = "COCKROACH_SKIP_FLAKY_TESTS"
> ```

Not sure if I understand the rationale for not using a flag. Could certainly create a flag here and then any package which imported `flaky` would see it, similar to what happens with logging. I'm fine with the env variable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-:-KUxDyLXSIDP_mchX7P_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[pkg/storage/client_metrics_test.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXFgU2Z1OjJC8jHVE:-KUvZK2_vRSn5-X4mP4O:bjftqeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_metrics_test.go#L152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to include these?
</details>
The comments? Yes.

---

*[pkg/util/flaky/flaky.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUxDfpi43Jvnoe_m0wh:-KUxmcFbhl0jqSCDXnIB:bqukr2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/util/flaky/flaky.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not sure if I understand the rationale for not using a flag. Could certainly create a flag here and then any package which imported `flaky` would see it, similar to what happens with logging. I'm fine with the env variable.
</details>
Any testing package would have to have this flag defined, meaning that I would have to add dummy imports everywhere that doesn't import `flaky` already (similarly to how we have dummy imports for our `log` package in a few places to get the `--verbosity` flag). Developer comfort is another plus for the env var (annoying to pass the flag every time).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[build/teamcity-acceptance.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgf-r2-4:-KUxnTO8p6lsfDoOnVM-:b-kistsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-acceptance.sh#L4)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Unfortunately, this won't work - `builder.sh` only passes a particular set of environment variables through to the Docker environment. You'll need to add `COCKROACH_SKIP_FLAKY_TESTS` to the list, or just change `builder.sh` to pass through all of `env`.
</details>
I replied via email, but it didn't attach. In this case, this is just fine, since the binary will be run in this shell, not the builder.

---

*[build/teamcity-stress.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgg-r2-4:-KUxnaRAKKYye4TmRysf:b-ppxvxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-stress.sh#L4)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> ditto
</details>
I replied via email, but it didn't attach. In this case, you need to add this variable to the `env` invocation below.

---

*[build/teamcity-test.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgh-r2-4:-KUxnenpSHiGGQuCLcJQ:bn69yw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-test.sh#L4)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> ditto
</details>
I replied via email, but it didn't attach. In this case, you need to add an `env` invocation below with this variable.

---

*[build/teamcity-testrace.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgi-r2-4:-KUxnjvzcoBG9f2m6weU:bn69yw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-testrace.sh#L4)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> ditto
</details>
I replied via email, but it didn't attach. In this case, you need to add an `env` invocation below with this variable.

---

*[pkg/storage/client_metrics_test.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXFgU2Z1OjJC8jHVE:-KUxnJ92dttr1J0-hSh7:b78j8hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_metrics_test.go#L152)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The comments? Yes.
</details>
And also the test changes. What do these have to do with the rest of the PR? Random cleanup?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[pkg/sql/copy_in_test.go, line 296 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvX92nEkgRRdB5VliK:-KUye7X-aQUAdS44Nlx4:b-2rq4ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/sql/copy_in_test.go#L296)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> cc @mjibson 
</details>
Augh. I forgot about this. I've created a cockroach issue to track it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197)*
<!-- Sent from Reviewable.io -->

tbg committed:
*: skip flaky tests only optionally

Using `t.Skip{f,}` as the weapon of choice for flaky tests is not a good
solution as tests which never run tend to rot and wind up not only flaky, but
completely broken (see #10141).

On the other hand, *not* skipping flaky tests is also doomed to fail since the
toll on developers is too high.

Introduce an alternative which skips tests based on the environment variable
`COCKROACH_SKIP_FLAKY_TESTS` whenever `flaky.Register(<issue_num>, <reason>)`
is called, along with a simple lint to avoid accidental use of `t.Skip{,f}`.
With the environment variable set to a truthy value, the test is skipped with
an informative message. Otherwise, a similarly informative message is logged as
a `Warning`, but the test continues to run.

Change existing `Skip{f,}`s to use this method, fixing up some tests in the
process and documenting completely broken ones (through an extra `Skip`).

Change the TeamCity build scripts to set `COCKROACH_SKIP_FLAKY_TESTS=true`.

The intention is that henceforth, we can more liberally mark tests as flaky and
still not have them rot as they still run occasionally, though unsatisfyingly
mostly on developer's machines. There's also a small upshot to stress testing:
there's no need to unskip a test which is to be stressed.

NB: Using a command-line parameter was considered, but the need to have that
flag registered with every package made it the less palatable option.

tschottdorf commented:
> I don't feel very strongly about this, but I'd like to hear more about the purpose.
> Seems to me that the problem is not that we don't run flaky tests, is that they stay skipped for too long. And I guess you're trying to address this by having them run on developer's machines, so that they'll feel some pain and ping the respective bugs, and that would prod the assigned person to fix them.
> Does that sum it? If so, how about we create some mechanism that more directly prods people to fix them. E.g. automated emails, or 30s in the team meeting where the people with the most/oldest skipped tests are publicly shamed? These would not have the downside that... running flaky tests is annoying when they fail on your unrelated change.

> I'm not sure about the policy either. I understand the problem, but should annoying all devs really be the preferred way to get these bugs fixed?

No, developer pain isn't a goal here. Rather it's to avoid that tests rot more while they are flaky. This is relatively rare but does happen, and now it will be noticed instantly (you could call this developer pain, but I wouldn't as it requires immediate action and then the pain is gone, even if it means Skipping one more test). Of course it's not unwelcome if you find that your own tests flake on you every once in a while when running `make test` before pushing a PR, and if that incentivizes a fix.

I agree that things stay skipped for too long, and I'm all for making us more proactive (unfortunately right now only some of us are really proactive), but the main goal is first to avoid tests rotting further, second incentivize growth of some of that infrastructure.

For example, you could imagine running flaky tests but not counting towards a failed build (hard because timeouts), but still posting to the issue. You could imagine better inventory of them (something like `guru -scope . -tags test callers util/flaky.go:#1370`, which unfortunately breaks but so does looking up the callers in my vim integration), ...

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[build/teamcity-acceptance.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgf-r2-4:-KV-HEUvJSslPcxqVNky:bgdt0d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-acceptance.sh#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I replied via email, but it didn't attach. In this case, this is just fine, since the binary will be run in this shell, not the builder.
</details>
Now passing this directly to the test invocation to avoid rot and confusion.

---

*[build/teamcity-stress.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgg-r2-4:-KV-H7sNpOQ03R4rdmQc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-stress.sh#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I replied via email, but it didn't attach. In this case, you need to add this variable to the `env` invocation below.
</details>
Done.

---

*[build/teamcity-test.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgh-r2-4:-KV-H9ZE4U-uOxb0A4Ex:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-test.sh#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I replied via email, but it didn't attach. In this case, you need to add an `env` invocation below with this variable.
</details>
Done.

---

*[build/teamcity-testrace.sh, line 4 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUv8HrS1kARaywAsCgi-r2-4:-KV-HAJ6cLKu1SMaA1y7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/build/teamcity-testrace.sh#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I replied via email, but it didn't attach. In this case, you need to add an `env` invocation below with this variable.
</details>
Done.

---

*[pkg/acceptance/cli_test.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvWesu7QpgvJjk5sej:-KV-HrKWKL96uCc6h3j9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/acceptance/cli_test.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is actually not flaky, it's something different
</details>
Done.

---

*[pkg/acceptance/reference_test.go, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvWr5om_Kl_BXq4HXw:-KV-HqdcEfhbLPb3mkqv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/acceptance/reference_test.go#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[pkg/storage/client_metrics_test.go, line 152 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXFgU2Z1OjJC8jHVE:-KV-G3scu4L-AFBPvLja:b-m4jf5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_metrics_test.go#L152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> And also the test changes. What do these have to do with the rest of the PR? Random cleanup?
</details>
Oh, sorry. `TestStoreMetrics` is the caller here, and this rotted while the test was skipped, requiring the changes.

---

*[pkg/storage/client_raft_test.go, line 2089 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXUgHqu8J1vBPi69m:-KV-GKHvMd2MDY29CNZ5:bi4afxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/storage/client_raft_test.go#L2089)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this number changed?
</details>
Whoops, restored.

---

*[pkg/util/flaky/flaky.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXzrvwk-3fto4LbbF:-KV-GN03aYhnUoE7kxLp:bgtblk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/util/flaky/flaky.go#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> any reason not to use `testing.TB` directly?
</details>
I like not pulling `testing` into non-testing packages (though of course the intention here is to never import this from any others).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-:-KV-HuhE3HknQUEJKyH6:b-9f60dh)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
IMO this should live under `testutils`. If we want to add another env var to enable longer running tests, it would probably belong in the same package - but then the package name `flaky` is (while admittedly awesome) too particular.

If we wanted to set up a daily stress run of flaky tests only (to get failure rates), how would we go about setting it up? The only way I can think of is doing a "normal" run and grepping for the "test is flaky" messages to get a list.

tamird commented:
Reviewed 4 of 23 files at r3, 23 of 23 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/util/flaky/flaky.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197#-KUvXzrvwk-3fto4LbbF:-KV0QGEWNpgv73rFPq6w:b-289uzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f1b0dab66353bb33d4dbd63d9c7cc2538700c6e/pkg/util/flaky/flaky.go#L27)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I like not pulling `testing` into non-testing packages (though of course the intention here is to never import this from any others).
</details>
I think @raduberinde suggested that this be in testutils, which might be the right thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10197)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I object to running flaky tests by default on developers' machines. This will have the effect of prioritizing tests based on how often they fail, rather than the priority of the issue they test for. For example, a test that flakes 1/3 of the time would be unacceptable for developer productivity (and many of our tests have had that level of flakiness when they were skipped). 

We already spend a lot of time fixing flaky tests; our flakiness problem is not because we ignore tests once they're skipped. The problem is that we *introduce* flaky tests at a significant rate; any general solution to flaky tests must address the source.

tschottdorf commented:
Abandoning, if anyone wants to pick this up it's all theirs. I'll leave the branch around on my fork.
